src/NuGet.Clients/NuGet.PackageManagement.UI/Actions/UIActionEngine.cs

```diff
@@ -311,22 +318,48 @@ private static string GenerateUpgradeReport(string projectName, string backupPat
             packageEnumerationTime.Start();
             try
             {
-                // collect the install state of the existing packages
-                foreach (IProjectContextInfo project in uiService.Projects)
+                IServiceBroker sb = uiService.UIContext.ServiceBroker;
+                if (!userAction.IsSolutionLevel && userAction.Action == NuGetProjectActionType.Install && uiService.Projects?.Count() == 1)
```

Comment:
From lines 322 to 365, this could be rewritten as follows to have less branches and be more concise. Please test this code though if you decide to go with it! :-)

```suggestion
                int projectsCount = uiService.Projects.Count();
                foreach (IProjectContextInfo project in uiService.Projects)
                {
                    IEnumerable<IPackageReferenceContextInfo> installedPackages = null;
                    if (projectsCount == 1 && !userAction.IsSolutionLevel && userAction.Action == NuGetProjectActionType.Install && project.ProjectStyle == ProjectModel.ProjectStyle.PackageReference && project.ProjectKind == NuGetProjectKind.PackageReference)
                    {
                        IInstalledAndTransitivePackages installedAndTransitives = await project.GetInstalledAndTransitivePackagesAsync(sb, cancellationToken);
                        installedPackages = installedAndTransitives.InstalledPackages;

                        packageToInstallWasTransitive = false;
                        string packageIdToInstall = VSTelemetryServiceUtility.NormalizePackageId(userAction.PackageId);
                        foreach (IPackageReferenceContextInfo transitivePackage in installedAndTransitives.TransitivePackages)
                        {
                            if (packageIdToInstall == VSTelemetryServiceUtility.NormalizePackageId(transitivePackage.Identity.Id))
                            {
                                packageToInstallWasTransitive = true;
                                break;
                            }
                        }
                    }
                    else
                    {
                        installedPackages = await project.GetInstalledPackagesAsync(sb, cancellationToken);
                    }

                    foreach (IPackageReferenceContextInfo package in installedPackages)
                    {
                        existingPackages.Add(CreatePackageTuple(package));
                    }
                }
```